### PR TITLE
Parallel Head

### PR DESCRIPTION
**Pull request by gabenavarro**
### Description

This PR introduces a **Parallel Head** capability for Evo2/Hyena models that enables **multi-task training** with independent, toggleable heads and loss weights for DNA LM (cross-entropy), RNA expression (MSE), and peptide-mapping (MSE).

**Key additions**

* `ParallelHeadTransform` (utils/heads/parallel_head.py)

  * Attaches one or more parallel heads to the underlying Hyena core:

    * `rna_seq_head`: TP-aware `ColumnParallelLinear(hidden_size → 1)`
    * `pep_map_head`: TP-aware `ColumnParallelLinear(hidden_size → 1)` (optional)
  * Adds per-head toggles and weights:

    * `parallel_dna`, `parallel_rna`, `parallel_pep`
    * `dna_loss_weight`, `rna_loss_weight`, `pep_loss_weight`
  * Detects and targets the correct **forward wrapper** vs **core** Hyena model and safely overrides `forward()` only on the forward-target model. Shares heads between wrapper and core.
  * Preserves shared-embedding weight tying for LM output when enabled.
  * Computes:

    * **DNA CE loss** with optional lowercase reweighting via `make_upper_case` + `reweighted_cross_entropy`.
    * **RNA / PEP losses** via masked element-wise **MSE** (supports `loss_mask` alignment and CP/PP/TP shapes).
  * Returns either a dict of logits in inference mode or a (bfloat16) total loss in training mode.
* `parallel_head_forward_step_fn(model, batch)`

  * Thin adapter to pass standard GPT inputs **plus** optional `rna_seq_targets` and `pep_map_targets`.
* `parallel_head_data_step_fn(dataloader_iter, use_mtp=False)`

  * Prepares a batch for PP/TP/CP: moves required tensors to the correct device/host sets and slices by context-parallel rank.
  * Ensures `tokens`, `position_ids`, `labels`, `loss_mask`, `attention_mask`, optional `cu_seqlens`, and optional RNA/PEP targets are routed to the expected stages.
* `utils/heads/__init__.py` export for the new transform.

**Why this matters**

* Enables **parallel, weighted multi-objective training** (e.g., sequence modeling + omics regression) without invasive changes to model internals.
* Keeps compatibility with NeMo/Megatron **PP/TP/CP** execution and shared weight tying.
* Creates a clear path to add more heads later (e.g., CHIP-seq, protein abundance, MEA features).

#### Usage

Minimal example: attach the transform and train with both DNA and RNA heads; optionally enable peptide mapping.

```python
import torch
from nemo.collections.llm.gpt.model.megatron.hyena.hyena_model import HyenaModel
from bionemo.evo2.utils.heads.parallel_head import (
    ParallelHeadTransform,
    parallel_head_forward_step_fn,
    parallel_head_data_step_fn,
)

# 1) Build or load your Hyena/Evo2 model as usual
model: HyenaModel = build_or_load_hyena_model_somehow()

# 2) Apply the transform (choose your weights/toggles)
transform = ParallelHeadTransform(
    dna_loss_weight=1.0,
    rna_loss_weight=1.0,
    pep_loss_weight=0.5,
    parallel_dna=True,
    parallel_rna=True,
    parallel_pep=False,  # set True to enable PEP map head
)
model = transform(model)

# 3) Training step (data step handles PP/TP/CP device routing)
dataloader_iter = iter(train_dataloader)

batch = parallel_head_data_step_fn(dataloader_iter)  # returns dict[str, Tensor] sliced per CP rank
# Expected keys (subset/optional): tokens, position_ids, attention_mask, labels, loss_mask,
# rna_seq_targets, pep_map_targets, cu_seqlens, cu_seqlens_argmin, max_seqlen

loss = parallel_head_forward_step_fn(model, batch)

loss.backward()
optimizer.step()
optimizer.zero_grad()
```

**Batch contract (shapes are per PP/CP/TP partition):**

* `tokens`: [S_local, B_local] or [B_local, S_local] depending on upstream; function handles CP slicing.
* `labels`: same layout as LM head expects.
* `loss_mask`: broadcastable to CE and MSE shapes; transposes handled when needed.
* `rna_seq_targets`, `pep_map_targets`: typically `[B, S]` (or transposed); auto-aligned with logits incl. transpose fixes.
* Inference (no `labels`): returns `{'logits', 'rna_seq_logits', 'pep_map_logits'}`; logits are gathered appropriately for LM when enabled.

### Type of changes

* [x] New feature (non-breaking change which adds functionality)
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] Refactor
* [ ] Documentation update
* [ ] Other (please describe):

### CI Pipeline Configuration

Suggested labels depending on how deep you want coverage:

* Add **`ciflow:all`** to run unit, slow, and notebook tests for bionemo2.
* If runtime is a concern, use **`ciflow:slow`** (single-GPU slow tests) or keep default unit tests.
* If examples/notebooks are added later for this feature, include **`ciflow:notebooks`**.

*Unit tests marked `@pytest.mark.multi_gpu` / `@pytest.mark.distributed` are excluded by default in PR CI.*

### Notes / Caveats (please sanity-check)

* In `parallel_head_data_step_fn`, the checks for peptide targets mix `_batch` vs `batch` in a couple places. For consistency and to avoid `NameError` or silently skipping device moves, consider changing:

  * `if "pep_map_targets" in batch:` → `if "pep_map_targets" in _batch:` (both occurrences)
* `total_loss` is converted to **bfloat16** at the end. That’s usually fine under bf16 training but double-check if your training step expects fp32 loss for scaler/logging.
* `reduction='none'` for RNA/PEP MSE returns element-wise losses; masking is applied but **you may want to `mean()` or `sum()` before returning** if your trainer expects a scalar. If you intend per-token/per-sample loss for later aggregation, leave as is; otherwise consider:

  ```python
  rna_seq_loss = (rna_seq_loss * aligned_loss_mask).sum() / (aligned_loss_mask.sum() + 1e-8)
  ```
* `gather_output=True` in TP heads is set so inference returns full-tensor logits. If you prefer sharded outputs and gather later, flip accordingly.
* The transform detects **core vs wrapper** via presence of `embedding/decoder/output_layer`; if a future model reorganizes these, you may need to update discovery heuristics.

### Documentation touchpoints (follow-ups)

* Add a short how-to in `docs` for multi-head training (batch contract, toggles, common shape pitfalls).
* Provide a toy test/dataset stub that sets `rna_seq_targets` to verify forward/backward.

### CODE REVIEW

**Comment by gabenavarro** [comment]
Sorry, wrong branch pull request merge!
